mentat/code_file_manager.py

```diff
@@ -126,106 +121,66 @@ def _add_file(self, abs_path: Path):
-                    self.file_lines[code_change.name] = self._read_file(
-                        code_change.name
+    # Mainly does checks on if file is in context, file exists, file is unchanged, etc.
+    def write_changes_to_files(self, file_edits: list[FileEdit]):
+        for file_edit in file_edits:
+            rel_path = Path(os.path.relpath(file_edit.file_path, self.config.git_root))
+            if file_edit.is_creation:
+                if file_edit.file_path.exists():
+                    raise MentatError(
+                        f"Model attempted to create file {file_edit.file_path} which"
+                        " already exists"
                     )
-                case _:
-                    file_changes[rel_path].append(code_change)
-
-        for rel_path, changes in file_changes.items():
-            abs_path = self.config.git_root / rel_path
-            new_code_lines = self._get_new_code_lines(rel_path, changes)
-            if new_code_lines:
-                if abs_path not in self.code_context.files:
+                self._add_file(file_edit.file_path)
+            else:
+                if file_edit.file_path not in self.code_context.files:
+                    raise MentatError(
+                        f"Attempted to edit file {file_edit.file_path} not in context"
+                    )
+                elif not file_edit.file_path.exists():
                     raise MentatError(
-                        f"Attempted to edit file {abs_path} not in context"
+                        f"Attempted to edit non-existent file {file_edit.file_path}"
                     )
-                with open(abs_path, "w") as f:
-                    f.write("\n".join(new_code_lines))
+
+            if file_edit.is_deletion:
+                cprint(f"Are you sure you want to delete {rel_path}?", "red")
+                if self.user_input_manager.ask_yes_no(default_yes=False):
+                    cprint(f"Deleting {rel_path}...", "red")
+                    self._delete_file(file_edit.file_path)
+                    continue
+                else:
+                    cprint(f"Not deleting {rel_path}", "green")
+
+            if not file_edit.is_creation:
+                stored_lines = self.file_lines[rel_path]
+                if stored_lines != self.read_file(rel_path):
+                    logging.info(
+                        f"File '{file_edit.file_path}' changed while generating changes"
+                    )
+                    cprint(
+                        f"File '{rel_path}' changed while generating; current"
+                        " file changes will be erased. Continue?",
+                        color="light_yellow",
+                    )
+                    if not self.user_input_manager.ask_yes_no(default_yes=False):
+                        cprint(f"Not applying changes to file {rel_path}")
+            else:
+                stored_lines = []
+
+            if file_edit.rename_file_path is not None:
```

Comment:
Hmm, I'm not entirely sure about this because that stages the rename; I don't think we want mentat to stage anything unless the user specifically asks us to; and of course when the user eventually stages the changes git will recognize the rename either way. It is nice having git recognize it immediately though. What do you think?

Reply:
hmm yeah we don't want to stage. I believe `git mv` forces git to recognize it as a rename (which it sometimes fails, if similarity isn't high enough). Let's leave as is for now so we don't stage. Perhaps in the future we could have Mentat "remember" a rename and so when it commits for the user it'll inform git